providers/providers.go

```diff
@@ -16,11 +16,16 @@ import (
 	k8serrors "k8s.io/apimachinery/pkg/util/errors"
 )
 
+const (
+	CodeChallengeMethodPlain = "plain"
+	CodeChallengeMethodS256  = "S256"
+)
+
 // Provider represents an upstream identity provider implementation
 type Provider interface {
 	Data() *ProviderData
-	GetLoginURL(redirectURI, finalRedirect string, nonce string) string
-	Redeem(ctx context.Context, redirectURI, code string) (*sessions.SessionState, error)
+	GetLoginURL(redirectURI, finalRedirect, nonce, codeChallenge, codeChallengeMethod string) string
```

Comment:
Now that both of these methods have multiple arguments that are essentially optional depending on the provider (used to be 1 with nonce only for OIDC). Should we refactor this to take in a Struct with default nilable versions of nonce, code challenge & method?

(This is probably more a question for @JoelSpeed)

Reply:
A recent merge added the ability for users to give their own login url parameters. I could refactor this a bit so that the code challenge parameters are added in `oauthproxy.go` to the `extraParams` argument as well if that is preferred to the longer function signatures. Let me know what you guys think.

Reply:
I think moving this into `extraParams` makes sense to keep the logic more centralised rather than spreading it through each of the different login URLs.

That said, does that then mean we will send it for providers that don't set it? (I think not because of the detection (eventually) meaning we only set it when a provider supports it)

Reply:
Can do in extra params. It would only be set now if you provide the command flag. And you're right eventually only set if the provider supports it